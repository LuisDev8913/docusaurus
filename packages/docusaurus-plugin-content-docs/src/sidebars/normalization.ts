/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import type {
  NormalizedSidebarItem,
  NormalizedSidebar,
  NormalizedSidebars,
  SidebarCategoriesShorthand,
  SidebarItemCategoryConfig,
  SidebarItemConfig,
  SidebarConfig,
  SidebarsConfig,
  NormalizedSidebarItemCategory,
} from './types';
import {isCategoriesShorthand} from './utils';
import _ from 'lodash';

function normalizeCategoriesShorthand(
  sidebar: SidebarCategoriesShorthand,
): SidebarItemCategoryConfig[] {
  return Object.entries(sidebar).map(([label, items]) => ({
    type: 'category',
    label,
    items,
  }));
}

/**
 * Normalizes recursively item and all its children. Ensures that at the end
 * each item will be an object with the corresponding type.
 */
export function normalizeItem(
  item: SidebarItemConfig,
): NormalizedSidebarItem[] {
  if (typeof item === 'string') {
    return [
      {
        type: 'doc',
        id: item,
      },
    ];
  }
  if (isCategoriesShorthand(item)) {
    return normalizeCategoriesShorthand(item).flatMap((subItem) =>
      normalizeItem(subItem),
    );
  }
  if (item.type === 'category') {
    if (typeof item.items !== 'undefined' && typeof item.items !== 'object') {
      throw new Error(
        `Invalid category ${JSON.stringify(
          item,
        )}: items must be an array of sidebar items or a category shorthand`,
      );
    }
    const normalizedCategory: NormalizedSidebarItemCategory = {
      ...item,
      items: Array.isArray(item.items)
        ? item.items.flatMap((subItem) => normalizeItem(subItem))
        : normalizeCategoriesShorthand(item.items).flatMap((subItem) =>
            normalizeItem(subItem),
          ),
    };
    return [normalizedCategory];
  }
  return [item];
}

function normalizeSidebar(sidebar: SidebarConfig): NormalizedSidebar {
  const normalizedSidebar = Array.isArray(sidebar)
    ? sidebar
    : normalizeCategoriesShorthand(sidebar);

  return normalizedSidebar.flatMap((subItem) => normalizeItem(subItem));
}

export function normalizeSidebars(
  sidebars: SidebarsConfig,
): NormalizedSidebars {
  return _.mapValues(sidebars, normalizeSidebar);
}
