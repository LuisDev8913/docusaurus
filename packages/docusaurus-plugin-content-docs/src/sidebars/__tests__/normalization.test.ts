/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import {normalizeSidebars} from '../normalization';

describe('normalization', () => {
  test('normalizes shorthands', () => {
    expect(
      normalizeSidebars({
        sidebar: {
          Category: ['doc1', 'doc2'],
          'Category 2': {
            'Subcategory 1': ['doc3', 'doc4'],
            'Subcategory 2': ['doc5', 'doc6'],
          },
        },
      }),
    ).toMatchSnapshot();

    expect(
      normalizeSidebars({
        sidebar: [
          {
            type: 'link',
            label: 'Google',
            href: 'https://google.com',
          },
          {
            'Category 1': ['doc1', 'doc2'],
            'Category 2': ['doc3', 'doc4'],
          },
        ],
      }),
    ).toMatchSnapshot();
  });
});
