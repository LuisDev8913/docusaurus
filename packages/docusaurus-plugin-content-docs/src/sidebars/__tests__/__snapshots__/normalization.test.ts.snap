// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`normalization normalizes shorthands 1`] = `
Object {
  "sidebar": Array [
    Object {
      "items": Array [
        Object {
          "id": "doc1",
          "type": "doc",
        },
        Object {
          "id": "doc2",
          "type": "doc",
        },
      ],
      "label": "Category",
      "type": "category",
    },
    Object {
      "items": Array [
        Object {
          "items": Array [
            Object {
              "id": "doc3",
              "type": "doc",
            },
            Object {
              "id": "doc4",
              "type": "doc",
            },
          ],
          "label": "Subcategory 1",
          "type": "category",
        },
        Object {
          "items": Array [
            Object {
              "id": "doc5",
              "type": "doc",
            },
            Object {
              "id": "doc6",
              "type": "doc",
            },
          ],
          "label": "Subcategory 2",
          "type": "category",
        },
      ],
      "label": "Category 2",
      "type": "category",
    },
  ],
}
`;

exports[`normalization normalizes shorthands 2`] = `
Object {
  "sidebar": Array [
    Object {
      "href": "https://google.com",
      "label": "Google",
      "type": "link",
    },
    Object {
      "items": Array [
        Object {
          "id": "doc1",
          "type": "doc",
        },
        Object {
          "id": "doc2",
          "type": "doc",
        },
      ],
      "label": "Category 1",
      "type": "category",
    },
    Object {
      "items": Array [
        Object {
          "id": "doc3",
          "type": "doc",
        },
        Object {
          "id": "doc4",
          "type": "doc",
        },
      ],
      "label": "Category 2",
      "type": "category",
    },
  ],
}
`;
